src/notifications/notificationsManager.ts

```diff
@@ -73,38 +53,99 @@ export class NotificationsManager {
-			newNotifications.push(notification);
+		// Resolve notifications
+		const notificationItems = new Map<string, NotificationTreeItem>();
+		await Promise.all(notificationsData.notifications.map(async notification => {
+			const cachedNotification = this._notifications.get(notification.key);
+			if (cachedNotification && cachedNotification.notification.updatedAd.getTime() === notification.updatedAd.getTime()) {
+				notificationItems.set(notification.key, cachedNotification);
+				return;
+			}
+
+			const model = await this._notificationProvider.getNotificationModel(notification);
+			if (!model) {
+				return;
+			}
+
+			notificationItems.set(notification.key, {
+				notification, model, kind: 'notification'
+			});
+		}));
+
+		// Calculate notification priority
+		if (this.sortingMethod === NotificationsSortMethod.Priority) {
+			const notificationsWithoutPriority = Array.from(notificationItems.values())
+				.filter(notification => notification.priority === undefined);
+
+			const notificationPriorities = await this._notificationProvider
+				.getNotificationsPriority(notificationsWithoutPriority);
+
+			for (const { key, priority, priorityReasoning } of notificationPriorities) {
+				const notification = notificationItems.get(key);
+				if (!notification) {
+					continue;
+				}
+
+				notification.priority = priority;
+				notification.priorityReason = priorityReasoning;
+
+				notificationItems.set(key, notification);
+			}
 		}
-		this._onDidChangeNotifications.fire(newNotifications);
+
+		this._notifications = notificationItems;
+		this._hasNextPage = notificationsData.hasNextPage;
+
+		const notifications = Array.from(this._notifications.values());
+		this._onDidChangeNotifications.fire(notifications);
+
+		return {
+			notifications: this._sortNotifications(notifications),
+			hasNextPage: this._hasNextPage
+		};
+	}
+
+	public getNotification(key: string): NotificationTreeItem | undefined {
+		return this._notifications.get(key);
 	}
 
-	public getAllNotifications(): INotificationItem[] {
+	public getAllNotifications(): NotificationTreeItem[] {
 		return Array.from(this._notifications.values());
 	}
 
-	public updateNotificationPriority(updates: NotificationUpdate[]) {
-		const updated: INotificationItem[] = [];
-		for (const update of updates) {
+	public async markAsRead(notificationIdentifier: { threadId: string, notificationKey: string }): Promise<void> {
+		await this._notificationProvider.markAsRead(notificationIdentifier);
 
-			const notification = this._notifications.get(update.key);
-			if (notification) {
-				notification.priority = update.priority;
-				notification.priorityReasoning = update.priorityReasoning;
-				updated.push(notification);
-			}
+		const notification = this._notifications.get(notificationIdentifier.notificationKey);
+		if (notification) {
+			this._onDidChangeNotifications.fire([notification]);
+			this._notifications.delete(notificationIdentifier.notificationKey);
```

Comment:
Might be better to do the `delete` before the `fire` to ensure that the state that is maintained in the `NotificationsManager` is fully up-to-date before event listeners are called.